File: 0287-find-the-duplicate-number/solution.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:

        # ==================================================
        #  Two Pointer                                     =
        # ==================================================
        # time  : O(n)
        # space : O(1)

        slowP, fastP = nums[0], nums[0]

        # Find the intersection
        while True:
            slowP = nums[slowP]
            fastP = nums[nums[fastP]]
            if slowP == fastP: break

        # Find the entrance of the cycle
        slowP = nums[0]
        while slowP != fastP:
            slowP = nums[slowP]
            fastP = nums[fastP]

        return slowP

File: 0287-find-the-duplicate-number/solution.java
class Solution {
    /**
     * @time  : O(n)
     * @space : O(1)
     */
     
    public int findDuplicate(int[] nums) {
        int slowP = nums[0], fastP = nums[0];
        
        while(true) {
            slowP = nums[slowP];
            fastP = nums[nums[fastP]];
            if(slowP == fastP) break;
        }
        
        slowP = nums[0];
        while(slowP != fastP) {
            slowP = nums[slowP];
            fastP = nums[fastP];
        }
        
        return slowP;
    }
}
